Refactoring: Removing axis parameter from scales

vagnermcj opened the pull request:
This PR addresses [issue #29349](https://github.com/matplotlib/matplotlib/issues/29349) by refactoring built-in scale classes to make the axis parameter optional. The goal is to support future removal of axis while maintaining backward compatibility with current usage.

To achieve this, a decorator was introduced to allow constructors to accept or ignore axis. This was applied across all built-in scales. The scale_factory() function was updated to handle scale classes that do or do not use axis, and register_scale() now issues a deprecation warning if the constructor still expects it.

These changes prepare the codebase for eventual deprecation of axis in scale initializers, while avoiding any breaking changes in current usage.

My only doubt is which initializers for the scales should have the parameter as "axis = None". If all of them should have it, then i have to change the way my decorator works because axis could change position in parameters index

timhoffm commented on lib/matplotlib/scale.py:
```diff
@@ -103,14 +105,60 @@ def limit_range_for_scale(self, vmin, vmax, minpos):
+    def wrapper(self, *args, **kwargs):
+        sig = inspect.signature(init_func)
+        params = list(sig.parameters.values())
+        if params and params[1].name == "axis":
+            return init_func(self, *args, **kwargs)
```
AFAICS this is always true for all our decorated `__init__` functions, so the decorator effectively does nothing.

I think the decorator must be phrased slightly differently: For any `__init__(axis, ...)`, we currently have calls `TheScale(axis, ...)`. We now want to additionally support leaving out the *axis* parameter `TheScale(...)`. Derived scales may currently do `super().__init__(None, ...)`. The decorator will allow them to switch to `super().__init__(...)`.

For all scales that have only a single positional paramters, we can achieve this simply by making *axis* optional: `__init__(axis=None, ...)`. But this does not work for `FuncScale`, `FuncScaleLog` and `LogitScale`. So we need the decorator anyway and can apply it to all scales. This will also make it easier to later add deprecation messages.

For simplicity we can assume that the axis parameter is always passed positionally. Searching GitHub (e.g. [this search](https://github.com/search?q=%2Fclass+%5Cw%2B%5C%28%28.%5C.%29%3FLogScale%5C%29%2F+language%3Apython+path%3A*.py+NOT+is%3Afork+NOT+path%3A**%2Fmatplotlib%2Fscale.py+NOT+path%3A**%2Fsite-packages%2F**&type=code) - most of the expression is to remove code copies) did not yield any keyword usage of axis. I.e. basically we can reduce the wrapper to 

```
def wrapper(self, *args, **kwargs):
    if args and isinstance(args[0], ScaleBase):
         init_func(self, *args, **kwargs)
    else:
         # inject None as axis parameter
         init_func(self, None, *args, **kwargs)
 ```

We keep the parameter always around and used to still support custom derived classes. Only when they have migrated to not provide the parameter, we'll remove it from all our classes.

timhoffm reviewed (commented):
Thanks for the PR. This is a bit tricky to get right in a backward-compatible and migration-easy way. I think we need to do it slightly differently.

vagnermcj commented:
Apologies for the sequence of commits — I wanted to ensure each feedback point was properly addressed and tested incrementally. Please let me know if this is the appropriate way to validate ruff, mypy-stubtest, and CI tests in general, or if you'd prefer a cleaner commit history or a different workflow.

I changed back the deprecation warning function, and changed the decorator as well. However, in the decorator i had many errors when trying to check  "if args and isinstance(args[0], ScaleBase):", so instead of using ScaleBase as a reference i changed to "if args and isinstance(args[0], mpl.axis.Axis):". Does it make sense?

Let me know if anything else needs to be refined or if the structure of the changes could be improved further. Thanks again for your time and guidance throughout the review!

timhoffm commented on lib/matplotlib/scale.pyi:
```diff
@@ -56,12 +57,12 @@ class LogScale(ScaleBase):
-        self,
-        axis: Axis | None,
+        self: LogScale,
+        axis: Union[Axis, None] = None,
```
```suggestion
        axis: Axis | None = ...,
```

`Axis | None` is equivalent to `Union[Axis, None]`, but the preferred spelling. You don't repeat default values in stub files. Instead, you place an ellipsis.

timhoffm commented on lib/matplotlib/scale.py:
```diff
@@ -103,14 +104,58 @@ def limit_range_for_scale(self, vmin, vmax, minpos):
 
-    def __init__(self, axis):
+    @handle_axis_parameter
+    def __init__(self, axis=None):
```
Is there an advantage in making `axis` optional here? That should be handled be handled by the decorator. If there's no particular reason, then I'd prefer to keep the signature formally unchanged for now.

timhoffm commented on lib/matplotlib/scale.py:
```diff
@@ -725,7 +776,12 @@ def scale_factory(scale, axis, **kwargs):
-    return scale_cls(axis, **kwargs)
+    try:
+        return scale_cls(axis, **kwargs)
+    except TypeError as e:
+        if 'unexpected keyword argument' in str(e) or 'positional argument' in str(e):
+            return scale_cls(**kwargs)
+        raise
```
Are you sure this is safe? Basing logic on the exception message only can be brittle. Minimally, I'd want an explicit comment on what we catch here. But I'd rather split this out into a separte PR as we need tests.

timhoffm reviewed (commented):
> Apologies for the sequence of commits — I wanted to ensure each feedback point was properly addressed and tested incrementally. Please let me know if this is the appropriate way to validate ruff, mypy-stubtest, and CI tests in general, or if you'd prefer a cleaner commit history or a different workflow.

It does not matter for small PRs. We'll typically review the full diff and squash-merge.

> However, in the decorator i had many errors when trying to check "if args and isinstance(args[0], ScaleBase):", so instead of using ScaleBase as a reference i changed to "if args and isinstance(args[0], mpl.axis.Axis):". Does it make sense?

Yes. Sorry my bad for the incorrect suggestion.

> Let me know if anything else needs to be refined or if the structure of the changes could be improved further. Thanks again for your time and guidance throughout the review!

You should add a test that the decorated scales can be created with and without *axis* passed. - Either test on one of the scales, or write an explicit test for the decorator.

vagnermcj commented:
Do i still need to make some change in the code? I see that it was accepted but some tests are failing 

It's my first PR so i'm not sure 😅

timhoffm commented:
We require two reviews from core developers. The second one is still pending. IIRC the test failures were unrelated, but since the test runs have been cleaned up, you could [rebase onto main](https://matplotlib.org/devdocs/devel/development_workflow.html#rebase-onto-upstream-main) and [force-push](https://matplotlib.org/devdocs/devel/development_workflow.html#push-with-force).

QuLogic commented on lib/matplotlib/scale.py:
```diff
@@ -279,7 +329,8 @@ class LogScale(ScaleBase):
 
-    def __init__(self, axis, *, base=10, subs=None, nonpositive="clip"):
+    @_make_axis_parameter_optional
+    def __init__(self, axis=None, *, base=10, subs=None, nonpositive="clip"):
```
Is there a reason this was changed from `axis` to `axis=None`, while _some_ of the others weren't?

QuLogic replied:
We normally cannot re-order parameters, as that would break callers, though the decorator may avoid that problem. Looking back at previous discussion, I see that @timhoffm kept these parameter/argument passing for derived classes, so I suppose we can't just delete it entirely and will have to leave them around for now.

QuLogic commented on lib/matplotlib/scale.py:
```diff
@@ -103,13 +104,61 @@ def limit_range_for_scale(self, vmin, vmax, minpos):
+        if args and isinstance(args[0], mpl.axis.Axis):
+            return init_func(self, *args, **kwargs)
+        else:
+            # Remove 'axis' from kwargs to avoid double assignment
+            kwargs.pop('axis', None)
+            return init_func(self, None, *args, **kwargs)
```
Shouldn't the keyword-argument version be passed as the `axis` if it's there?
```suggestion
            # Remove 'axis' from kwargs to avoid double assignment
            return init_func(self, kwargs.pop('axis', None), *args, **kwargs)
```
